### PR TITLE
[LOOP-1160] Support manual glucose entry on the bolus screen

### PR DESCRIPTION
**Pull request by mpangburn**
- When glucose data is stale, offer a prompt for the user to enter a manual glucose value
- If glucose data returns while the user is entering a glucose value, retract the entry field and offer a new recommendation

Requires https://github.com/tidepool-org/LoopKit/pull/182

### CODE REVIEW

File: Loop/Managers/LoopDataManager.swift
Comment (mpangburn) [inline Loop/Managers/LoopDataManager.swift:977]: While changing the bolus screen over the last few iterations, it's been a challenge to decouple LoopDataManager's use stored data (and effects derived from stored data) from the computations which combine them into a prediction or recommended bolus.

Adding API here to 'override' the stored effects with ones computed elsewhere isn't _super_ clean, but it's a minimal change, and significant refactoring in algorithm-land is high risk.

File: Loop/Managers/LoopDataManager.swift
Comment (mpangburn) [inline Loop/Managers/LoopDataManager.swift:1113]: Glucose effects aren't computed when glucose data is stale, since no prediction is vended. To offer a prediction from manual glucose, these effects are computed locally on demand.

File: Loop/Managers/LoopDataManager.swift
Comment (mpangburn) [inline Loop/Managers/LoopDataManager.swift:1176]: Momentum isn't relevant since CGM data is absent; RC _could_ be relevant if glucose data is < 30min old, though last I spoke with Product the ask was to use only insulin and carbs for this prediction.

File: Loop/Managers/LoopDataManager.swift
Comment (mpangburn) [inline Loop/Managers/LoopDataManager.swift:1207]: This is a split of one function into two to decouple validation based on stored glucose data.

**Review by mpangburn — COMMENTED** [review]
A few notes for reviewers:

File: Loop/Managers/LoopDataManager.swift
Comment (mpangburn) [inline Loop/Managers/LoopDataManager.swift:1254]: This split out decouples validation logic based on stored glucose data

File: Loop/Extensions/UIFont.swift
Comment (darinkrauss) [inline Loop/Extensions/UIFont.swift:31]: [nit] Any way to put this on UIFont, non-static, so instead of `.heavy(.title1)` it would be `.title1.heavy`?

File: Loop/Managers/LoopDataManager.swift
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift]: [nit] Should this have an "Override" suffix, too, so that it is clear it is overriding the normal behavior?

File: Loop/Views/BolusEntryView.swift
Comment (darinkrauss) [inline Loop/Views/BolusEntryView.swift:419]: [nit] Does this include units (e.g. "mg/dL")? If not, should it?

File: Loop/View Models/BolusEntryViewModel.swift
Comment (darinkrauss) [inline Loop/View Models/BolusEntryViewModel.swift]: [nit] This is inconsistent with above code that uses lots of `self?`

File: Loop/View Models/BolusEntryViewModel.swift
Comment (darinkrauss) [inline Loop/View Models/BolusEntryViewModel.swift:527]: [nit] Consider `if let manualGlucoseSample = DispatchQueue.main.sync { self.manualGlucoseSample } {` to remove bang `!` below.

File: Loop/View Models/BolusEntryViewModel.swift
Comment (darinkrauss) [inline Loop/View Models/BolusEntryViewModel.swift]: [nit] Here you use `self.` prefix, but not the matching two lines above.

File: Loop/Managers/LoopDataManager.swift
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift:1009]: Why these changes (from line 1009-1063)? I understand the need to provide an "override" above, but (to my untrained eye) this look like it is changing the behavior other than providing overrides.

File: Loop/Managers/LoopDataManager.swift
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift:1128]: [nit] Consider checking `effectCalculationError` and throwing here to avoid unnecessary work below.

File: Loop/Managers/LoopDataManager.swift
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift:1103]: Would we ever be replacing a carb entry with a manual glucose?

File: Loop/Managers/LoopDataManager.swift
Comment (darinkrauss) [reply Loop/Managers/LoopDataManager.swift:1009]: I cannot discern from the code changes what the for these changes was.

File: Loop/Managers/LoopDataManager.swift
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift:1116]: So, `includingPendingInsulin` will be `true` if there is any positive pending temp basal or bolus (from `recommendBolusForManualGlucose`. Why would we exclude glucose effects from doses if there wasn't any pending insulin? Wouldn't the glucose effect from no insulin be zero?

Admittedly, my knowledge of the algorithm is very weak, so I probably just don't understand what I'm seeing here.

**Review by darinkrauss — COMMENTED** [review]
I have a few comments and a number of questions to help me understand this PR. Admittedly, my algorithm knowledge is pretty weak at this point. I should bite the bullet and really try to understand all of the intricacies. Anyways, just ask for another review once you've responded and I'll take another look. Thanks!

File: Loop/Extensions/UIFont.swift
Comment (mpangburn) [reply Loop/Extensions/UIFont.swift:31]: It would be tricky to finagle, since `title1` isn't a UIFont but a UIFont.TextStyle, and implicit member expressions can't be chains. So making it a member on UIFont would look something like `UIFont.preferredFont(forTextStyle: .title1).heavy()`.

File: Loop/Views/BolusEntryView.swift
Comment (mpangburn) [reply Loop/Views/BolusEntryView.swift:419]: Yes, QuantityFormatter's `string(from:for:)` includes both value and unit.

File: Loop/View Models/BolusEntryViewModel.swift
Comment (mpangburn) [reply Loop/View Models/BolusEntryViewModel.swift:527]: Trailing closures can't be adjacent to the start of blocks, thus `sync`'s argument label becomes required, resulting in 
```swift
if let manualGlucoseSample = DispatchQueue.main.sync(execute: { self.manualGlucoseSample }) {
    // ...
} 
```
which started to feel long, but I'm not opposed to it if you prefer it.

File: Loop/Managers/LoopDataManager.swift
Comment (mpangburn) [reply Loop/Managers/LoopDataManager.swift:1103]: Editing a carb entry brings the user to the bolus screen right after; if glucose data is missing, they'll have the option to enter a manual glucose at that point.

File: Loop/Managers/LoopDataManager.swift
Comment (mpangburn) [reply Loop/Managers/LoopDataManager.swift:1116]: This is subtle; pending insulin can be negative. If negative pending insulin is included when computing a bolus recommendation, the recommendation may be compensating for the future withholding of insulin, resulting in a system that's competing against itself. This is a change we made when incorporating pending insulin directly into the prediction.

File: Loop/Managers/LoopDataManager.swift
Comment (mpangburn) [reply Loop/Managers/LoopDataManager.swift:1009]: Ah, I forgot to leave a comment here; thanks for calling this out. 

I had managed to contrive a scenario where, in the absence of historical carb data and glucose data, `recentCarbEntries` was `nil` even though `potentialCarbEntry` was non-`nil`, which (from the branch that occurred on 1006 prior to this commit) resulted in the carb effect of the potential carb entry not being included in the prediction.

This group of changes ensures that, regardless of whether `carbEffect` or `recentCarbEntries` is `nil`, the carb effect of `potentialCarbEntry` is always computed and included. The bottom lines make an analogous change for `insulinEffect` and `potentialBolus`.

Comment (mpangburn) [reply Loop/Managers/LoopDataManager.swift:1009]: This circumstance only manifested during manual glucose entry, since in other contexts this function would only be run while loop data was valid, which meant that `recentCarbEntries` was always assigned (even if just an empty array).

**Comment by mpangburn** [comment]
@darinkrauss Hopefully these answers help give context, happy to answer any other questions that come up. I really appreciate your careful eye here.

File: Loop/Extensions/UIFont.swift
Comment (darinkrauss) [reply Loop/Extensions/UIFont.swift:31]: All good, then. Just thought I'd ask. 👍

File: Loop/Managers/LoopDataManager.swift
Comment (darinkrauss) [reply Loop/Managers/LoopDataManager.swift:1103]: Ah, right! Thanks!

File: Loop/Managers/LoopDataManager.swift
Comment (darinkrauss) [reply Loop/Managers/LoopDataManager.swift:1116]: Ah, I see my confusion, `getPendingInsulin` does not count negative pending temp basal.

I think my confusion stems from the fact that `getPendingInsulin` isn't really returning pending insulin. Furthermore, since this function is only *ever* compared against > 0, the better long term choice might be to rename `getPendingInsulin` to `shouldlncludePendingInsulin` and include the comparison (> 0) within the function.

Just a thought...

File: Loop/Managers/LoopDataManager.swift
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift:1053]: [nit] Consider putting lines 1053-1058 inside of block starting at line 1060 (`if inputs.contains(.insulin)`). The `computationInsulinEffect` is only used inside that block and it is unnecessary work if `.insulin` is not included in the `inputs`.

Comment (darinkrauss) [reply Loop/Managers/LoopDataManager.swift:1053]: In fact, I think you might be able to simplify this block with lines 1061-1063.

File: Loop/Managers/LoopDataManager.swift
Comment (darinkrauss) [reply Loop/Managers/LoopDataManager.swift:1009]: I think this will need approval from @ps2. I don't have enough understanding at this point.

**Review by darinkrauss — APPROVED** [review]
LGTM! Approving with the caveat that @ps2 needs to approve the changes to `predictGlucose` as I don't have enough internal knowledge to do so.

File: Loop/Managers/LoopDataManager.swift
Comment (ps2) [reply Loop/Managers/LoopDataManager.swift:977]: Being worried about refactoring is likely evidence of lack of tests. Can we add some tests for this code, then?

**Review by ps2 — COMMENTED** [review]
LGTM! I do think that LDM is expanding and growing in complexity, and could really use some unit tests.

File: Loop/Managers/LoopDataManager.swift
Comment (mpangburn) [reply Loop/Managers/LoopDataManager.swift:977]: It's a catch-22:
- Being able to refactor safely requires unit tests.
- Writing unit tests requires that code be sufficiently factored to break dependencies.

I'll see if I can pull together some tests here.

**Review by ps2 — APPROVED** [review]
I do think we need to add testing at some point, but for timing's sake we need to get this into design review and QA sooner than later.  This looks great. Thanks Michael!

**Pull request merged**
